DietPi-PREP | Fix DietPi system detection; Disable Jessie; More enhancements

MichaIng opened the pull request:
**Status**: Ready
- [x] Go in checking code from ENV prep step until end 😉

**Commit list/description**:
+ DietPi-PREP | Resolve DietPi system being always detected due to syntax issue
+ DietPi-PREP | Remove Jessie install option from DietPi-PREP
+ DietPi-PREP | Remove "pre-patch_file" as well from downloaded code
+ DietPi-PREP | Move changelog to /boot as well; Prefix it and readme with "dietpi-"
+ DietPi-PREP | Create initial DietPi RAMdisk with 10M, which matches the final value as well and is more than sufficient
+ DietPi-PREP | Move last remaining pre-req APT package and log dir creation out of first step and removed it
+ DietPi-PREP | Remove Wheezy from supported initial distro versions. We cannot guarantee that 2 dist-upgrades to min target Stretch will work well
+ DietPi-PREP | Do not remove /etc/logrotate.d to preserve package specific rotation rules, when user chooses related logging mode
+ DietPi-PREP | Clean bash and nano history + PREP script for all users
+ DietPi-PREP | Further coding: Check file/folder existence and remove with full output then, to allow error output, e.g. due to FS corruption
+ CHANGELOG | Add info about disabling fresh Jessie images
+ DietPi-PREP | Simplify auto kernel package detection. "apt-cache search" succeeds for and "apt-get install" marks local packages as well. Merge "linux-image" and "linux-dtb" detection.
+ DietPi-PREP | Remove all conflicting APT sources at top of script, before running first APT update
+ DietPi-PREP | Use "rm -f" to avoid "missing files" errors on wildcard removal, instead of checking via "ls -A"
+ DietPi-PREP | Save PREP script location before cd to /tmp, to allow it's removal at end of script
+ DietPi-PREP | Minor coding, wording and visual

MichaIng committed:
v6.20

+ DietPi-PREP | Resolve DietPi system being always detected due to syntax issue
+ DietPi-PREP | Remove Jessie install option from DietPi-PREP
+ DietPi-PREP | Remove "pre-patch_file" as well from downloaded code
+ DietPi-PREP | Move changelog to /boot as well; Prefix it and readme with "dietpi-"
+ DietPi-PREP | Create initial DietPi RAMdisk with 10M, which matches the final value as well and is more than sufficient
+ DietPi-PREP | Move last remaining pre-req APT package and log dir creation out of first step and removed it
+ DietPi-PREP | Prefer checking file/dir existence before removal, instead of hiding output, which allows critical errors (e.g. I/O error due to FS corruption) to show up.
+ DietPi-PREP | Minor visual and coding enhancements

MichaIng committed:
v6.20

+ DietPi-PREP | Remove Wheezy from supported initial distro versions. We cannot guarantee that 2 dist-upgrades to min target Stretch will work well
+ DietPi-PREP | Do not remove /etc/logrotate.d to preserve package specific rotation rules, when user chooses related logging mode
+ DietPi-PREP | Clean bash and nano history + PREP script for all users
+ DietPi-PREP | Further coding: Check file/folder existence and remove with full output then, to allow error output, e.g. due to FS corruption

MichaIng committed:
v6.20

+ CHANGELOG | Add info about disabling fresh Jessie images

MichaIng committed:
v6.20

+ DietPi-PREP | Minor

MichaIng committed:
v6.20

+ DietPi-PREP | Syntax
+ DietPi-PREP | Allow /var/tmp/dietpi/logs/dietpi-ramlog_store removal
+ DietPi-PREP | Umount /DietPi, before removal
+ DietPi-PREP | Avoid doubled /var/tmp/dietpi/logs/dietpi-ramlog_store removal

MichaIng commented:
Following issues on my VM:
- [x] We need to re-add ForceIPv4 by default, again ran into the issue that APT and wget attempted to access via IPv6 to some sources, but hang on this forever. This also fits to: `CONFIG_PREFER_IPV4=1` being our default, but currently applied on 1st run, but not yet on PREP.
- DNS resolving issue on the testing VM, where I also tested PREP on netman controlled system and such. `resolvconf` not pre-installed. After `G_AGA`:
```
The following packages will be REMOVED:
  busybox* dconf-gsettings-backend* dconf-service* dropbear*
  dropbear-initramfs* glib-networking* glib-networking-common*
  glib-networking-services* gsettings-desktop-schemas* iproute2*
  libbluetooth3* libdbus-1-3* libdconf1* libelf1* libexpat1* libglib2.0-0*
  libgudev-1.0-0* libicu57* libjansson4* libmm-glib0* libmnl0* libndp0*
  libnl-3-200* libnl-genl-3-200* libnm0* libpcsclite1* libpolkit-agent-1-0*
  libpolkit-backend-1-0* libpolkit-gobject-1-0* libproxy1v5* libsoup2.4-1*
  libssl1.0.2* libteamdctl0* libxml2* wpasupplicant*
```
then `G_AGDUG`:
```
Err:1 https://deb.debian.org/debian stretch/main amd64 base-files amd64 9.9+deb9u6
  Could not resolve host: deb.debian.org
Err:2 https://deb.debian.org/debian-security stretch/updates/main amd64 perl-base amd64 5.24.1-3+deb9u5
  Could not resolve host: deb.debian.org
Err:3 https://deb.debian.org/debian stretch/main amd64 systemd-sysv amd64 232-25+deb9u6
  Could not resolve host: deb.debian.org
...
```
Reason is empty `/etc/resolv.conf` and no remaining package taking care of this. Manually adding `nameserver` entry allows to go on. So we could `echo 'nameserver 8.8.8.8' > /etc/resolv.conf` to assure the DNS resolving. Since `resolvconf` is installed later, if not yet done, on next ifup, the entry will be reset.

But actually, as can be seen above, there is a general issue when doing `G_AGA` before installing our required packages, is that some dependency packages are first autoremoved and then reinstalled again. This looks like an unnecessary disk I/O, in this case it was `iproute2` + it's dependencies.

So actually I would prefer that we install our requirements first, then do `G_AGA && G_AGDUG` afterwards. This also resolves the DNS issue with temporary missing resolver. Also it makes the code easier: No separate G_AGI for critical kernel/firmware packages required, just add everything to the array and do a single G_AGI.
@Fourdee 
What do you think?

MichaIng committed:
v6.20

+ DietPi-PREP | Re-add preferring IPv4 by default to workaround hanging access attempts in some cases and match dietpi.txt default as well

Fourdee commented:
@MichaIng 

> So actually I would prefer that we install our requirements first, then do G_AGA && G_AGDUG afterwards.

Yep, sounds good 👍

Fourdee commented on PREP_SYSTEM_FOR_DIETPI.sh:
```diff
@@ -1595,17 +1592,23 @@ _EOF_
 
 		# - Remove PREP script
-		rm /root/PREP_SYSTEM_FOR_DIETPI.sh &> /dev/null
+		[[ -f /root/PREP_SYSTEM_FOR_DIETPI.sh ]] && rm /root/PREP_SYSTEM_FOR_DIETPI.sh
```
Might be easier to simply?:
```
rm $(find / -name 'PREP_SYSTEM_FOR_DIETPI.sh') &> /dev/null
```

MichaIng replied:
@Fourdee 
I just though that find might take a long time. On my RPi at least it does, of course depending on SDcard as well. And, if I remember right, it does search all `/mnt/*` drives as well, which increases the time much more. For our image creation not a big problem, if we have no external drives attached, not much data on disk etc., but for end users with perhaps already attached drives and custom data on disk, it properly leads to an unnecessary delay?

On the other hand, we don't know where the user downloaded PREP to, so only find will surely find it. But user home dirs are best non-RAM location guesses 😉.... Ah wait, isn't there a bash variable to find current script location? mom..

`$0` or `$BASH_SOURCE`?
________
Okay not as easy, since most bash variables contain relative paths only. What I found is:
`"$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)/$(basename "${BASH_SOURCE[0]}")"`: https://stackoverflow.com/a/17123575
- ${BASH_SOURCE[0]} works for sourced scripts as well.
- Only with `cd` it allows via `pwd` to show full path instead of relative one.

But: This does not work, if you call the script with relative path, e.g. `./PREP, then` navigate to another location within the script, since ${BASH_SOURCE[0]} still contains only relative path, which is `./` then and leads to `cd` stay at current location...

I think it is by design, that from within a script you should not be able to identify it's full path without proper permissions?

It looks like any `cd` destroys the ability to identify the original full script path, if called via related path, e.g. `./script` or `../path/to/script` and such 🤔.

Fourdee reviewed (approved):
@MichaIng 

Good stuff 👍  please merge when required.

MichaIng committed:
Merge branch 'dev' into PREP

MichaIng commented:
Okay, new approach to delete full dir content, including all kind of hidden files, excluding `.` and `..`
```
2018-12-15 18:43:37 root@micha:/tmp# l test
total 0
drwxr-xr-x 2 root root 40 Dec 15 18:43 .d
-rw-r--r-- 1 root root  0 Dec 15 18:43 .t
-rw-r--r-- 1 root root  0 Dec 15 18:42 test
-rw-r--r-- 1 root root  0 Dec 15 18:42 .test
-rw-r--r-- 1 root root  0 Dec 15 18:42 ..test
drwxr-xr-x 2 root root 40 Dec 15 18:42 testdir
drwxr-xr-x 2 root root 40 Dec 15 18:42 .testdir
drwxr-xr-x 2 root root 40 Dec 15 18:42 ..testdir
2018-12-15 18:43:38 root@micha:/tmp# rm -Rf test/{,.??,.[^.]}*
2018-12-15 18:44:01 root@micha:/tmp# l test
total 0
2018-12-15 18:44:03 root@micha:/tmp# rm -Rf test/{,.??,.[^.]}*
2018-12-15 18:44:15 root@micha:/tmp#
```
- Problem is that `rm *` does not match leading dot (hidden) files.
- `rm .*` removes hidden files, but also matches `.` and `..`, throwing an error, luckily, because we of course don't want the dir and parent dir to  be removed:
```
2018-12-15 18:47:38 root@micha:/tmp/test/test/test# rm -Rf .*
rm: refusing to remove '.' or '..' directory: skipping '.'
rm: refusing to remove '.' or '..' directory: skipping '..'
```
- `rm .??*` is the usual way to exclude `.` and `..` by matching only 3 char files.
- `rm .[^]` removes as well 2 char files, but excludes `..`.
- `rm -f` does not throw an error, if no match found, but still shows other kind of errors, e.g. I/O, missing permission or as seen above refuse to remove `.` and `..`. So this actually prevents the need use `ls -A` to check wildcard file/dir existence and `[[ -f file ]]` as well. Although, the latter check is around 200x faster than `rm -f` attempt, so if we want to remove a large amount of possibly existing files/dirs, the check could be still beneficial for performance reasons...although, we are talking about ~6ms for one rm -f on my RPi2... To avoid long lines/code beauty, I think this can be ignored.

MichaIng committed:
v6.20

+ DietPi-PREP | Simplify auto kernel package detection. "apt-cache search" succeeds for and "apt-get install" marks local packages as well. Merge "linux-image" and "linux-dtb" detection.
+ DietPi-PREP | Remove all conflicting APT sources at top of script, before running first APT update
+ DietPi-PREP | Use "rm -f" to avoid "missing files" errors on wildcard removal, instead of checking via "ls -A"
+ DietPi-PREP | Save PREP script location before cd to /tmp, to allow it's removal at end of script
+ DietPi-PREP | Minor coding and wording

MichaIng committed:
v6.20

+ DietPi-PREP | Fix ugly whiptail before dietpi-globals load on Dropbear connection
+ DietPi-PREP | Minor coding, output and wording

MichaIng committed:
v6.20

+ DietPi-Patch | Remove possible PREP leftover "pre-patch_file"
+ DietPi-Patch | Rename "README.md" to "dietpi-README.md" to stand out from other distro/kernel/firmware related readme and licence files

MichaIng committed:
v6.20

+ DietPi-RAMlog | Remove hidden files from disk location as well, before copying new logs

MichaIng committed:
v6.20

+ DietPi-PREP | Fix kernel pkg output, due to "--get-selections" => "-l" switch, which shows more details about the installed kernel version

MichaIng commented:
Will merge for further testing, and do the APT install/autoremove switch in separate PR for separate review/testing.

MichaIng commented:
@Fourdee 
Not sure, if it's related to this PR, but on first boot (directly after applying DietPi-PREP), terminal login prompt appears, before even `preboot` has finished. When logging in, one sees the login banner telling that system prep has not yet finished, preboot, boot, postboot finishing, but of course no `dietpi-software` autostart, since login (script) did already run. Manual `/DietPi/dietpi/login` is then required.

Never had this before, maybe we need to add `Before=` login target to our boot scripts?
